File: data structures and algorithms/python/08_023.py
def min_team_members(intervals, K):
   # Sort intervals by start time
   intervals.sort(key=lambda x: x[0])
   
   # Keep track of covered intervals
   covered = [False] * len(intervals)
   team_count = 0
   
   while not all(covered):
      team_count += 1
      
      # Find the earliest uncovered interval
      earliest_uncovered = None
      for i in range(len(intervals)):
         if not covered[i]:
            earliest_uncovered = i
            break
      
      if earliest_uncovered is None:
         break
         
      # Start shift at the beginning of the earliest uncovered
      shift_start = intervals[earliest_uncovered][0]
      shift_end = shift_start + K
      
      # Cover all intervals that fit within this shift
      for i in range(len(intervals)):
         if (not covered[i] and 
             intervals[i][0] >= shift_start and 
             intervals[i][1] <= shift_end):
            covered[i] = True
   
   return team_count

# Test with example
intervals = [[1, 3], [2, 5], [6, 8], [8, 10], [11, 12]]
K = 5
print(min_team_members(intervals, K))  # Output: 2

File: data structures and algorithms/cs/08_023.cs
using System;
using System.Linq;

public class IntervalCoverageOptimization
{
   public static int MinTeamMembers(int[][] intervals,
                                    int K)
   {
      Array.Sort(intervals, (a, b) =>
         a[0].CompareTo(b[0]));

      bool[] covered = new bool[intervals.Length];
      int teamCount = 0;

      while (!AllCovered(covered))
      {
         teamCount++;

         int earliestUncovered = -1;
         for (int i = 0; i < intervals.Length; i++)
         {
            if (!covered[i])
            {
               earliestUncovered = i;
               break;
            }
         }

         if (earliestUncovered == -1)
         {
            break;
         }

         int shiftStart = intervals[earliestUncovered][0];
         int shiftEnd = shiftStart + K;

         for (int i = 0; intervals.Length; i++)
         {
            if (!covered[i] && intervals[i][0] >=
                shiftStart && intervals[i][1] <=
                shiftEnd)
            {
               covered[i] = true;
            }
         }
      }

      return teamCount;
   }

   private static bool AllCovered(bool[] covered)
   {
      return covered.All(c => c);
   }

   public static void Main(string[] args)
   {
      int[][] intervals = {
            new int[] { 1, 3 },
            new int[] { 2, 5 },
            new int[] { 6, 8 },
            new int[] { 8, 10 },
            new int[] { 11, 12 }
        };
      int K = 5;
      Console.WriteLine("Minimum team members needed: "
                        + MinTeamMembers(intervals, K));
      // Output: 2
   }
}